stdlib/builtins.pyi

```diff
@@ -940,23 +939,33 @@ class bool(int):
     def __invert__(self) -> int: ...
 
 @final
-class slice(Generic[_StartT, _StopT, _StepT]):
+class slice(Generic[_StartT_co, _StopT_co, _StepT_co]):
     @property
-    def start(self) -> _StartT: ...
+    def start(self) -> _StartT_co: ...
     @property
-    def step(self) -> _StepT: ...
+    def step(self) -> _StepT_co: ...
     @property
-    def stop(self) -> _StopT: ...
+    def stop(self) -> _StopT_co: ...
+    # generic slice --------------------------------------------------------------------
     @overload
-    def __new__(cls, stop: int | None, /) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
+    def __new__(cls, start: None, stop: None = None, step: None = None, /) -> slice[Any, Any, Any]: ...
+    # unary overloads ------------------------------------------------------------------
     @overload
-    def __new__(
-        cls, start: int | None, stop: int | None, step: int | None = None, /
-    ) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
+    def __new__(cls, stop: _T2, /) -> slice[_T2 | Any, _T2, Any]: ...
+    # binary overloads -----------------------------------------------------------------
+    @overload
+    def __new__(cls, start: _T1, stop: None, step: None = None, /) -> slice[_T1, _T1 | Any, Any]: ...
+    @overload
+    def __new__(cls, start: None, stop: _T2, step: None = None, /) -> slice[_T2 | Any, _T2, Any]: ...
+    @overload
+    def __new__(cls, start: _T1, stop: _T2, step: None = None, /) -> slice[_T1, _T2, Any]: ...
+    # ternary overloads ----------------------------------------------------------------
+    @overload
+    def __new__(cls, start: None, stop: None, step: _T3, /) -> slice[Any, Any, _T3]: ...
```

Comment:
Similarly for these three, I wonder if we could do this? It seems reasonable that if the type checker is assuming that `start` and `stop` are the same type, it should also assume that `step` is the same type as well (unless you give it explicit type annotations)

```suggestion
    @overload
    def __new__(cls, start: _T1, stop: None, step: None = None, /) -> slice[_T1, _T1 | Any, _T1 | Any]: ...
    @overload
    def __new__(cls, start: None, stop: _T2, step: None = None, /) -> slice[_T2 | Any, _T2, _T2 | Any]: ...
    @overload
    def __new__(cls, start: _T1, stop: _T2, step: None = None, /) -> slice[_T1, _T2, Any]: ...
    # ternary overloads ----------------------------------------------------------------
    @overload
    def __new__(cls, start: None, stop: None, step: _T3, /) -> slice[_T3 | Any, _T3 | Any, _T3]: ...
```

Reply:
Again, the counter-example I have in mind here is datetime/timedelta types.

Having `step` be the same type as `start` or `stop` only seems reasonable to me if and only if start and stop come from a [Linearly ordered group](https://en.wikipedia.org/wiki/Linearly_ordered_group).

This is true for integers and floats, but not for datetime/timedelta. Also, I don't see how it makes any sense for string types.

Reply:
`datetime`/`timedelta` is a very good counter-example here, thanks. Could you add some assertions to the test-case file involving `datetime` and `timedelta` so that we can be sure that these use cases won't break in the future?

Reply:
@randolf-scholz, I tried adding some tests for you, but I couldn't find anything involving datetimes that would fail with the change I'm proposing here. This is the test I tried (plus a few variations on it), but it still passes with my proposed change here:

```diff
--- a/stdlib/@tests/test_cases/builtins/check_slice.py
+++ b/stdlib/@tests/test_cases/builtins/check_slice.py
@@ -1,5 +1,5 @@
 """
-- The type hint `slice` should be compatible with the **all slices**:
+- The type hint `slice` should be compatible with *all slices*:
     - `slice(None)`, `slice(None, None)` and `slice(None, None, None)`. (⟿ `slice[?, ?, ?]`)
 - The type hint `slice[T]` should be compatible with:
     - `slice(None)`, `slice(None, None)` and `slice(None, None, None)` (⟿ `slice[?, ?, ?]`)
@@ -22,6 +22,9 @@
     - `slice(x, y, z)`  (⟿ `slice[X, Y, Z]`)
 """
 
+from __future__ import annotations
+
+import datetime as dt
 from typing import Any, Union
 from typing_extensions import assert_type
 
@@ -113,3 +116,20 @@ assert_type(slice(1, None, None).start, int)
 
 assert_type(slice(None, None, 1).step, int)
 # endregion Tests for slice properties -------------------------------------------------
+
+# region Integration tests for slices with datetimes -----------------------------------
+
+def requires_timedelta_step(x: slice[dt.datetime | None, dt.datetime | None, dt.timedelta]) -> None:
+    pass
+
+
+def foo(flag: bool, datetime: dt.datetime, delta: dt.timedelta) -> None:
+    if flag:
+        x = slice(datetime, None, delta)
+    else:
+        x = slice(None, datetime, delta)
+
+    requires_timedelta_step(x)
+
+
+# endregion Integration tests for slices with datetimes -----------------------------------
```

Do you have a test in mind that would fail with this proposed change? If so, could you add it to this PR?

If you don't have such a test in mind, I may merge anyway, but please let me know where we are with this.

Reply:
I added some tests for combining `__geitem__` with string / datetime slices. This is for example used in pandas:

https://pandas.pydata.org/docs/user_guide/timeseries.html#partial-string-indexing

Reply:
Seems like there are some problems with mypy. Also, I considered a different test:

```python
# test slices as a return type
def foo(flag: bool, value: DT) -> "slice[DT, None] | slice[None, DT]":
    if flag:
        return slice(value, None)  # slice[DT, DT|Any, Any] incompatible
    else:
        return slice(None, value)  # slice[DT|Any, DT, Any] incompatible
```

Similar to yours, it fails due to how the overloads are specified, but I would say this rather unusual slice usage. If we want it to pass we'd have to revert the `T|Any` back to plain `Any`.

Reply:
Thank you!



> Similar to yours, it fails due to how the overloads are specified, but I would say this rather unusual slice usage. If we want it to pass we'd have to revert the `T|Any` back to plain `Any`.

In the interests of getting this PR merged (which I think is a strict improvement on the status quo), I'm happy with either removing the test or changing the `T` to `T|Any`. I'll let you decide which to go with.

Reply:
> I'm happy with either removing the test or changing the T to T|Any

Just to double-check that there is no misunderstanding: the change would have to be to revert the `T | Any` back to plain `Any`.

This can be expressed as a consistency criterion:
    - `slice(x)` must be compatible with `slice[None, X, None]`, even if X is None.
    - `slice(x, y)` must be compatible with `slice[X,Y,None]`, even if X is None or Y is None.
    - `slice(x, y, z)` must be compatible with `slice[X, Y, Z]`, even if X, Y, or Z are `None`.

These three properties are incompatible with the`_T1 | Any` annotations in `__new__`.
Personally, I am in favor of this, following the principle that code that is formally correct shouldn't require `typing.cast` or `type: ignore`.

Reply:
Do you know if there is a simple way to statically verify/assert that two types are incompatible?
For instance, one would reorganize these tests to have a list of valid and invalid assigns for each call signature.

```pyhton
# checking slice(x)
x1: slice[int] = slice(1)  # OK
x1: slice[int, int] = slice(1)  # OK
x2: slice[None, int] = slice(1)  # OK
x3: slice[int, None] = slice(1)  # NOT OK
x4: slice[int | None] = slice(1)  # OK
```